js/reference.js

```diff
@@ -1114,30 +1114,80 @@ var ERMrest = (function(module) {
                     allNewData.push(newData);
 
                     submissionData[i] = {};
-                    for (var key in newData) {
-                        // if the key is part of the shortest key for the entity, the data needs to be aliased
+
+                    for (var keyIndex = 0; keyIndex < shortestKeyNames.length; keyIndex++) {
+                        var shortestKey = shortestKeyNames[keyIndex];
+
+                        // shortest key should always be aliased in case that key value was changed
                         // use a suffix of '_o' to represent changes to a value that's in the shortest key that was changed, everything else gets '_n'
-                        submissionData[i][key + oldAlias] = oldData[key];
-                        submissionData[i][key + newAlias] = newData[key];
+                        submissionData[i][shortestKey + oldAlias] = oldData[shortestKey];
+                        submissionData[i][shortestKey + newAlias] = newData[shortestKey];
+
+                        // don't add the current key to the column projections if it's already in there
+                        // this can happen when there are multiple tuples or when a key is part of the visible column set
+                        if (columnProjections.indexOf(shortestKey) === -1) {
+                            // keys are aliased and included in both the keyset and column projections set
+                            columnProjections.push(shortestKey);
+                        }
                     }
-                }
 
-                // The list of column names to use in the uri
-                columnProjections = Object.keys(tuples[0].data);
+                    // Loop through the visible columns so the submission data is based off of the visible columns list
+                    for (var m = 0; m < this.columns.length; m++) {
+                        var column = this.columns[m];
+                        var key;
+
+                        // if the column is disabled (generated or immutable), continue
+                        // no need to submit update data in the submission object
+                        if (column.inputDisabled) {
+                            continue;
+                        }
+
+                        if (this.columns[m].isPseudo) {
```

Comment:
Since you are going to change the code based on my comment below, change `this.columns[m]` to `column`. To make it consistent with the previous statement. (This is not a very important change but can help the readability of the code)